repos/os/src/drivers/spi/imx8q/ecspi/ecspi_driver.cc

```diff
@@ -0,0 +1,280 @@
+ * \date   2021-04-20
+ */
+
+/*
+ * Copyright (C) 2013-2021 Genode Labs GmbH
+ * Copyright (C) 2021 gapfruit AG
+ *
+ * This file is part of the Genode OS framework, which is distributed
+ * under the terms of the GNU Affero General Public License version 3.
+ */
+
+/* Local include */
+#include "ecspi_driver.h"
+
+/******************
+ ** Irq Handling **
+ *****************/
+
+void Spi::Ecspi_driver::_irq_handle()
+{
+
+	if (_mmio.read<Mmio::Status::Tx_fifo_full>()) {
+		if (_config.verbose) {
+			Genode::log("Tx FIFO is full.");
+		}
+	}
+
+	if (_mmio.read<Mmio::Status::Rx_fifo_full>()) {
+		if (_config.verbose) {
+			Genode::log("Rx FIFO is full.");
+		}
+	}
+
+	/* we don't want that to happen */
+	if (_mmio.read<Mmio::Status::Rx_fifo_overflow>()) {
+		Genode::error("Rx FIFO overflow, data loss!");
+		/* rx_overflow must be cleared. Reset value = 1 */
+		_mmio.write<Mmio::Status::Rx_fifo_overflow>(1);
+	}
+
+	if (_mmio.read<Mmio::Status::Rx_fifo_ready>()) {
+		if (_config.verbose) {
+			Genode::log("Rx FIFO is ready.");
+		}
+	}
+
+	if (_mmio.read<Mmio::Status::Tx_complete>()) {
+		if (_config.verbose) {
+			Genode::log("Tx transaction completed.");
+		}
+		/* tx_complete must be cleared. Reset value = 1 */
+		_mmio.write<Mmio::Status::Tx_complete>(1);
+		_sem_exchange.up();
+	}
+
+	_irq.ack_irq();
+}
+
+
+/******************
+ * Bus operations *
+ ******************/
+
+void Spi::Ecspi_driver::_bus_enable(Settings const &settings, size_t slave_select)
+{
+	/* enable SPI chipset, note on imx8q only ss: 0 is available */
+	_mmio.write<Mmio::Control::Enable>(1);
+
+	/* set master mode on the selected slave */
+	_mmio.write<Mmio::Control::Channel_select>(slave_select);
+	_mmio.write<Mmio::Control::Channel_mode>( 1 << slave_select);
+
+	/* configure loopback mode if needed */
+	if (_config.loopback) {
+		_mmio.write<Mmio::Test::Loop_back_ctl>(1);
+		/* use full FIFO size */
+		_mmio.write<Mmio::Test::Rx_fifo_counter>(0xFF);
+		_mmio.write<Mmio::Test::Tx_fifo_counter>(0xFF);
+	}
+
+	/* configure clock divider and line */
+	_mmio.write<Mmio::Control::Pre_divider>(_config.clock_divider >> 4);
+	_mmio.write<Mmio::Control::Post_divider>(_config.clock_divider & 0xF);
+	_mmio.write<Mmio::Config::Clock_phase>((settings.mode & 0x1) << slave_select);
+	_mmio.write<Mmio::Config::Clock_polarity>((settings.mode & 0x2) << slave_select);
+	_mmio.write<Mmio::Config::Clock_idle_state>(settings.clock_idle_state << slave_select);
+
+	/* configure data lines idle state */
+	_mmio.write<Mmio::Config::Data_idle_state>(settings.data_lines_idle_state << slave_select);
+
+	/* configure slave select active state */
+	_mmio.write<Mmio::Config::Slave_select_polarity>(settings.ss_line_active_state << slave_select);
+
+	/* enable interrupt */
+	_mmio.write<Mmio::Interrupt::Tx_completed_enable>(1);
+}
+
+
+void Spi::Ecspi_driver::_bus_disable()
+{
+	_mmio.write<Mmio::Control>(0);
+}
+
+
+void Spi::Ecspi_driver::_bus_execute_transaction()
+{
+	_sem_exchange.down();
```

Comment:
I think it is not necessary, but on an initial review, it had been suggested that I added it to make sure no data race occur.  
The `_sem_exchange` is flipped by the IRQ handler while the driver's ep thread is block by `_env.ep().wait_and_dispatch_one_io_signal();`. So if it is right to you, I can replace it by an integer?

Reply:
While reviewing, I discovered the use of this semaphore, which made me wondering about the requirement for multi-threading. Then, I did not find multiple threads being used and had the urge to ask ;-)

So in my opinion, it would be better to not use synchronization primitives in cases where only state/logic should be expressed. This makes the implementation more clear and also spares unnecessary overhead.